### PR TITLE
approval-voting: processed wakeups can also update approval state

### PR DESCRIPTION
**Pull request by rphmeier**
Related to #3847: a no-show that is pre-covered should now be marked as approved in the DB.

A short summary of changes: I renamed `import_checked_approval` to `advance_approval_state` so we could reuse the same code for more purposes than just approval checking imports. I added a new kind of state transition which is "processed wakeup". I changed the `process_wakeup` function to invoke `advance_approval_state` on every call. This means that if anything happens during regular wakeup processing to make the candidate approved, that it will be handled in the DB properly instead of dropped as it was before.

I added a test which I can confirm fails on `master` and passes in this branch.

### CODE REVIEW

File: node/core/approval-voting/src/lib.rs [hunk 1]
```diff
@@ -1865,13 +1867,15 @@ fn import_checked_approval(
 			status.required_tranches,
 		));
 
-		// We have no need to write the candidate entry if
+		// We have no need to write the candidate entry if all of the following
+		// are trueL
```
Comment (ordian) [inline node/core/approval-voting/src/lib.rs:1871]:
```suggestion
		// is true
```

File: node/core/approval-voting/src/tests.rs [hunk 2]
```diff
@@ -2555,3 +2555,197 @@ fn subsystem_assignment_not_triggered_if_at_maximum_but_clock_is_before_with_dri
...
+	// A, B are tranche 0.
+	// C is tranche 1.
+	//
+	// All assignments imported at once, and B, C approvals imported immediately.
+	// A no-shows, leading to being covered by C.
+	// Technically, this is an approved block, but it will be approved
+	// when the no-show timer hits, not as a response to an approval vote.
+	//
+	// Note that we have 6 validators, otherwise the 2nd approval triggers
+	// the >1/3 insta-approval condition.
```
Comment (ordian) [inline node/core/approval-voting/src/tests.rs:2570]: :heart: Thanks for the detailed comments!

**Review by ordian — APPROVED** [review]
Looks reasonable.

**Review by drahnr — APPROVED** [review]
LGTM, thanks for the explain-session :)

**Comment by rphmeier** [comment]
bot merge

**Comment by ordian** [comment]
bot merge

**Pull request merged**